### PR TITLE
Disable Net5CompatFileStream by default

### PR DESCRIPTION
**Pull request by adamsitnik**
* [x] ensure all tests are passing in runtime repo including the Outerloop tests
* [x] fix all bugs (#50424, #50277)
* [x] fix all breaking changes in runtime repo (#50367)
* [x] ensure all tests are passing in a few other repos:
   * [x] ASP.NET (https://github.com/dotnet/aspnetcore/pull/31441)
   * [x]  WinForms (https://github.com/dotnet/winforms/pull/4756) 
   * [x] SDK (https://github.com/dotnet/sdk/pull/16684)
* [x] document breaking changes (#50858, #50860)
* [x] rename `Legacy` to `Net5Compat` (#50470)

### CODE REVIEW

File: src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public static void LegacySwitchIsHonored()
                     .GetField("_strategy", BindingFlags.NonPublic | BindingFlags.Instance)
                     .GetValue(fileStream);
 
-                Assert.DoesNotContain(strategy.GetType().FullName, "Legacy");
+                Assert.Contains(strategy.GetType().FullName, "Legacy");
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs:23]: Can we also change the "Legacy" naming?

Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs:23]: > Can we also change the "Legacy" naming?

Of the internal types and properties? Sure, do you have any good suggestions? `Net5CompatFileStreamStrategy`? `BackwardCompatibileFileStreamStrategy`?

Comment (stephentoub) [reply src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs:23]: Yes,  they show up in exception stack traces, for example. Your first suggestion sounds good.

Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs:23]: > Your first suggestion sounds good.

Ok, I've added the rename to the TODO list of this PR. 

@carlossanlop @Jozkee please let me know if you have a better idea for a new name

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ private static bool GetLegacyFileStreamSetting()
 
             string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM");
             return envVar is null
-                ? true // legacy is currently enabled by default;
+                ? false // legacy is disabled by default;
                 : bool.IsTrueStringIgnoreCase(envVar) || envVar.Equals("1");
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs:23]: nit: 
```cs
return envVar != null && bool.IsTrueStringIgnoreCase(envVar) || envVar.Equals("1");

File: src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs (hunk 1 above)
Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/LegacyTests/LegacySwitchTests.cs]: I've sent #50470 with the rename

**Comment by adamsitnik** [comment]
All ASP.NET and WinForms tests are passing with the new strategy enabled: 
* https://github.com/dotnet/aspnetcore/pull/31441
* https://github.com/dotnet/winforms/pull/4756

Once the SDK build (https://github.com/dotnet/sdk/pull/16684) gets green as well I am going to provide the breaking changes doc in this PR and ask Team for the review. I am offline for the next four days, so this is going to happen on Tuesday and should allow us for shipping it as part of Preview 4.

**Comment by adamsitnik** [comment]
SDK is green https://github.com/dotnet/sdk/pull/16684

**Comment by adamsitnik** [comment]
@carlossanlop @Jozkee @stephentoub everything is ready (please see the PR description for full details), PTAL

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs [hunk 3]
```diff
@@ -11,17 +11,7 @@ internal static partial class FileStreamHelpers
...
-            {
-                return fileConfig;
-            }
-
-            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM");
-            return envVar is null
-                ? true // Net5Compat is currently enabled by default;
-                : bool.IsTrueStringIgnoreCase(envVar) || envVar.Equals("1");
-        }
+            => AppContextConfigHelper.GetBooleanConfig("System.IO.UseNet5CompatFileStream", "DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM");
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs:14]: Is this helper method still needed?
```C#
internal static bool UseNet5CompatStrategy { get; } =
    AppContextConfigHelper.GetBooleanConfig("System.IO.UseNet5CompatFileStream", "DOTNET_SYSTEM_IO_USENET5COMPATFILESTREAM");
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs (hunk 3 above)
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs:14]: Yes as it's reused by globalization as well:

https://github.com/dotnet/runtime/blob/6ef4b2e7aba70c514d85c2b43eac1616216bea55/src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs#L11-L18

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs:14]: By helper, I mean GetNet5CompatFileStreamSetting().

Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs]: > By helper, I mean GetNet5CompatFileStreamSetting()

It's not, I've just pushed a commit that removes it.

**Review by jozkee — APPROVED** [review]

**Pull request merged**